src/metabase/driver/sql_jdbc/actions.clj

```diff
@@ -341,56 +396,88 @@
-    (for [pk-value-map pk-value-maps]
-      (into [:and]
-            (for [[pk-name value] pk-value-map]
-              [:=
-               [:field (get pk-name->id pk-name) nil]
-               value])))))
-
 (defmethod actions/perform-action!* [:sql-jdbc :bulk/delete]
-  [driver _action {database-id :id, :as database} {:keys [table-id pk-value-maps]}]
-  (log/tracef "Deleting %d rows" (count pk-value-maps))
-  (with-jdbc-transaction [conn database-id]
-    (transduce
-      (m/indexed)
-      (fn
-        ([]
-         [[] []])
-        ([[errors]]
-         ;; if there were any errors throw an Exception with all the error messages.
-         (when (seq errors)
-           (.rollback conn)
-           (throw (ex-info (tru "Error(s) deleting rows.")
-                           {:status-code 400, :errors errors})))
-         ;; if there we no errors then return the created rows.
-         {:success true})
-        ([[errors] [row-index delete-filter]]
-         (try
-           (do-nested-transaction
-             driver
-             conn
-             (fn []
-               (actions/perform-action!*
-                 driver
-                 :row/delete
-                 database
-                 {:database   database-id
-                  :type       :query
-                  :query      {:source-table table-id
-                               :filter delete-filter}})))
-           [errors]
-           (catch Throwable e
-             [(conj errors {:index row-index, :error (ex-message e)})]))))
-      (pk-value-maps->honeysql-filter-clauses table-id pk-value-maps))))
+  [driver _action {database-id :id, :as database} {:keys [table-id rows]}]
+  (log/tracef "Deleting %d rows" (count rows))
+  (let [pk-name->id (table-id->pk-field-name->id table-id)]
+    ;; validate the keys in `rows`
+    (check-consistent-row-keys rows)
+    (check-rows-have-expected-columns-and-no-other-keys rows (keys pk-name->id))
+    (check-unique-rows rows)
+    ;; now do one `:row/delete` for each row
+    (perform-bulk-action-with-repeated-single-row-actions!
+     {:driver   driver
+      :database database
+      :action   :row/delete
+      :rows     rows
+      :xform    (comp (map (fn [row]
+                             {:database database-id
+                              :type     :query
+                              :query    {:source-table table-id
+                                         :filter       (row->mbql-filter-clause pk-name->id row)}}))
+                      #(completing % (fn [[errors _successes]]
+                                       (when (seq errors)
+                                         (throw (ex-info (tru "Error(s) deleting rows.")
+                                                         {:status-code 400, :errors errors})))
+                                       ;; `:bulk/delete` just returns a simple status message on success.
+                                       {:success true})))})))
```

Comment:
@kulyk pointed this out, this will probably be the best chance to make this consistent.

> 
>  hey, noticed a minor API inconsistency:
> after successful bulk delete, the BE responds with { "success": true }
> after successful bulk update, the BE responds with { "rows-updated": <int> }
> for the reference, row update/deletes also return a number of rows affected
> This isn’t a problem for the FE at all, but do we maybe want to send rows-deleted: <int> for /api/action/bulk/delete for consistency?
> There’s a bit more to that:
> bulk update returns something like { "rows-updated": 4 }
> row update returns something like { "rows-updated": [1] }
> row delete returns something like { "rows-deleted": [1] }
> Notice it’s an int for bulk update and it’s an array for row update/delete

Reply:
Don't think it makes sense to make a change to `bulk/delete` in the PR to _implement_ `bulk/update` . Let's just do it as a follow-on